--- a/net/minecraft/client/gui/screens/packs/PackSelectionModel.java
+++ b/net/minecraft/client/gui/screens/packs/PackSelectionModel.java
@@ -108,6 +_,10 @@
         boolean canMoveUp();
 
         boolean canMoveDown();
+
+        default boolean notHidden() {
+            return true;
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
@@ -207,6 +_,11 @@
         @Override
         public void moveDown() {
             this.move(1);
+        }
+
+        @Override
+        public boolean notHidden() {
+            return !this.pack.isHidden();
         }
     }
 
