--- a/net/minecraft/client/resources/model/ModelDiscovery.java
+++ b/net/minecraft/client/resources/model/ModelDiscovery.java
@@ -37,7 +_,8 @@
     private static Set<ModelResourceLocation> listMandatoryModels() {
         Set<ModelResourceLocation> set = new HashSet<>();
         BuiltInRegistries.ITEM.listElements().forEach(p_368638_ -> {
-            ResourceLocation resourcelocation = p_368638_.value().components().get(DataComponents.ITEM_MODEL);
+            // Forge: Force use vanilla components without gathering mod components in the event
+            ResourceLocation resourcelocation = p_368638_.value().vanillaComponents().get(DataComponents.ITEM_MODEL);
             if (resourcelocation != null) {
                 set.add(ModelResourceLocation.inventory(resourcelocation));
             }
