--- a/net/minecraft/client/resources/model/SimpleBakedModel.java
+++ b/net/minecraft/client/resources/model/SimpleBakedModel.java
@@ -29,7 +_,26 @@
     private final boolean usesBlockLight;
     private final TextureAtlasSprite particleIcon;
     private final ItemTransforms transforms;
-
+    /** Forge: Block render types to be used with {@linkplain net.minecraft.client.GraphicsStatus#FANCY fancy graphics} */
+    protected final net.minecraftforge.client.ChunkRenderTypeSet blockRenderTypes;
+    /** Forge: Block render types to be used with {@linkplain net.minecraft.client.GraphicsStatus#FAST fast graphics} */
+    protected final net.minecraftforge.client.ChunkRenderTypeSet blockRenderTypesFast;
+    /** Forge: If this model's {@linkplain #blockRenderTypes block render types} are rendering cutout, to account for older leaves model JSONs */
+    protected final boolean isRenderingCutout;
+
+    /** @deprecated Forge: Use {@linkplain #SimpleBakedModel(List, Map, boolean, boolean, boolean, TextureAtlasSprite, ItemTransforms, net.minecraftforge.client.RenderTypeGroup, net.minecraftforge.client.RenderTypeGroup) variant with RenderTypeGroup} **/
+    @Deprecated
+    public SimpleBakedModel(List<BakedQuad> p_119489_, Map<Direction, List<BakedQuad>> p_119490_, boolean p_119491_, boolean p_119492_, boolean p_119493_, TextureAtlasSprite p_119494_, ItemTransforms p_119495_) {
+        this(p_119489_, p_119490_, p_119491_, p_119492_, p_119493_, p_119494_, p_119495_, net.minecraftforge.client.RenderTypeGroup.EMPTY);
+    }
+
+    /** @deprecated Forge: Use {@linkplain #SimpleBakedModel(List, Map, boolean, boolean, boolean, TextureAtlasSprite, ItemTransforms, net.minecraftforge.client.RenderTypeGroup, net.minecraftforge.client.RenderTypeGroup) variant with RenderTypeGroup for fast graphics} **/
+    @Deprecated(forRemoval = true, since = "1.21.4")
+    public SimpleBakedModel(List<BakedQuad> p_119489_, Map<Direction, List<BakedQuad>> p_119490_, boolean p_119491_, boolean p_119492_, boolean p_119493_, TextureAtlasSprite p_119494_, ItemTransforms p_119495_, net.minecraftforge.client.RenderTypeGroup renderTypes) {
+        this(p_119489_, p_119490_, p_119491_, p_119492_, p_119493_, p_119494_, p_119495_, renderTypes, net.minecraftforge.client.RenderTypeGroup.EMPTY);
+    }
+
+    /** Constructor with {@link net.minecraftforge.client.RenderTypeGroup RenderTypeGroup} for fancy and fast graphics. Preferred over {@link net.minecraft.client.renderer.ItemBlockRenderTypes#setRenderLayer(net.minecraft.world.level.block.Block, net.minecraft.client.renderer.RenderType) ItemBlockRenderTypes.setRenderLayer(Block, RenderType)}. */
     public SimpleBakedModel(
         List<BakedQuad> p_119489_,
         Map<Direction, List<BakedQuad>> p_119490_,
@@ -37,7 +_,9 @@
         boolean p_119492_,
         boolean p_119493_,
         TextureAtlasSprite p_119494_,
-        ItemTransforms p_119495_
+        ItemTransforms p_119495_,
+        net.minecraftforge.client.RenderTypeGroup renderTypes,
+        net.minecraftforge.client.RenderTypeGroup renderTypesFast
     ) {
         this.unculledFaces = p_119489_;
         this.culledFaces = p_119490_;
@@ -46,6 +_,20 @@
         this.usesBlockLight = p_119492_;
         this.particleIcon = p_119494_;
         this.transforms = p_119495_;
+
+        boolean hasRenderTypes = renderTypes != null && !renderTypes.isEmpty();
+        boolean hasRenderTypesFast = renderTypesFast != null && !renderTypesFast.isEmpty();
+        this.blockRenderTypes = hasRenderTypes ? net.minecraftforge.client.ChunkRenderTypeSet.of(renderTypes.block()) : null;
+        this.blockRenderTypesFast = hasRenderTypesFast ? net.minecraftforge.client.ChunkRenderTypeSet.of(renderTypesFast.block()) : null;
+        this.isRenderingCutout = hasRenderTypes && (renderTypes.block() == net.minecraft.client.renderer.RenderType.cutout() || renderTypes.block() == net.minecraft.client.renderer.RenderType.cutoutMipped());
+    }
+
+    public static BakedModel bakeElements(List<BlockElement> p_377425_, TextureSlots p_378525_, SpriteGetter p_375793_, ModelState p_376680_, boolean p_375745_, boolean p_376866_, boolean p_376846_, ItemTransforms p_376883_) {
+        return bakeElements(p_377425_, p_378525_, p_375793_, p_376680_, p_375745_, p_376866_, p_376846_, p_376883_, null);
+    }
+
+    public static BakedModel bakeElements(List<BlockElement> p_377425_, TextureSlots p_378525_, SpriteGetter p_375793_, ModelState p_376680_, boolean p_375745_, boolean p_376866_, boolean p_376846_, ItemTransforms p_376883_, @Nullable net.minecraftforge.client.RenderTypeGroup renderType) {
+        return bakeElements(p_377425_, p_378525_, p_375793_, p_376680_, p_375745_, p_376866_, p_376846_, p_376883_, renderType, null);
     }
 
     public static BakedModel bakeElements(
@@ -56,7 +_,9 @@
         boolean p_375745_,
         boolean p_376866_,
         boolean p_376846_,
-        ItemTransforms p_376883_
+        ItemTransforms p_376883_,
+        @Nullable net.minecraftforge.client.RenderTypeGroup renderType,
+        @Nullable net.minecraftforge.client.RenderTypeGroup renderTypeFast
     ) {
         TextureAtlasSprite textureatlassprite = findSprite(p_375793_, p_378525_, "particle");
         SimpleBakedModel.Builder simplebakedmodel$builder = new SimpleBakedModel.Builder(p_375745_, p_376866_, p_376846_, p_376883_)
@@ -77,6 +_,13 @@
             }
         }
 
+        if (renderType != null && !renderType.isEmpty()) {
+            if (renderTypeFast != null && !renderTypeFast.isEmpty())
+                simplebakedmodel$builder.renderTypes(renderType, renderTypeFast);
+            else
+                simplebakedmodel$builder.renderTypes(renderType);
+        }
+
         return simplebakedmodel$builder.build();
     }
 
@@ -123,6 +_,19 @@
         return this.transforms;
     }
 
+    private static final net.minecraftforge.client.ChunkRenderTypeSet SOLID_BLOCK = net.minecraftforge.client.ChunkRenderTypeSet.of(net.minecraft.client.renderer.RenderType.solid());
+
+    @Override
+    public net.minecraftforge.client.ChunkRenderTypeSet getRenderTypes(BlockState state, RandomSource rand, net.minecraftforge.client.model.data.ModelData data) {
+        if (!net.minecraft.client.renderer.ItemBlockRenderTypes.isFancy()) {
+            if (blockRenderTypesFast != null)
+                return blockRenderTypesFast;
+            if (isRenderingCutout && state.getBlock() instanceof net.minecraft.world.level.block.LeavesBlock)
+                return SOLID_BLOCK;
+        }
+        return blockRenderTypes != null ? blockRenderTypes : BakedModel.super.getRenderTypes(state, rand, data);
+    }
+
     @OnlyIn(Dist.CLIENT)
     public static class Builder {
         private final ImmutableList.Builder<BakedQuad> unculledFaces = ImmutableList.builder();
@@ -164,13 +_,50 @@
             return this;
         }
 
+        private net.minecraftforge.client.RenderTypeGroup renderTypes = net.minecraftforge.client.RenderTypeGroup.EMPTY;
+        private net.minecraftforge.client.RenderTypeGroup renderTypesFast = net.minecraftforge.client.RenderTypeGroup.EMPTY;
+
+        /**
+         * Sets the render type to be used for this model, and will be used for any graphics setting.
+         * <p>
+         * If you need to set a specific render type for
+         * {@linkplain net.minecraft.client.GraphicsStatus#FAST fast graphics}, consider using
+         * {@link #renderTypes(net.minecraftforge.client.RenderTypeGroup, net.minecraftforge.client.RenderTypeGroup)
+         * renderTypes(RenderTypeGroup, RenderTypeGroup)} instead which allows choosing render types for both fancy and
+         * fast graphics.
+         *
+         * @apiNote If this model is for {@linkplain net.minecraft.world.level.block.LeavesBlock leaves} and if the
+         * given render type is either {@linkplain net.minecraft.client.renderer.RenderType#cutout() cutout} or
+         * {@linkplain net.minecraft.client.renderer.RenderType#cutoutMipped() cutout mipped}, it will be overridden
+         * with {@linkplain net.minecraft.client.renderer.RenderType#solid() solid} when fast graphics is enabled.
+         * @see #renderTypes(net.minecraftforge.client.RenderTypeGroup, net.minecraftforge.client.RenderTypeGroup)
+         * renderTypes(RenderTypeGroup, RenderTypeGroup)
+         */
+        public SimpleBakedModel.Builder renderTypes(net.minecraftforge.client.RenderTypeGroup renderTypes) {
+            return this.renderTypes(renderTypes, net.minecraftforge.client.RenderTypeGroup.EMPTY);
+        }
+
+        /**
+         * Sets the render types to be used for this model: one for
+         * {@linkplain net.minecraft.client.GraphicsStatus#FANCY fancy graphics} and one for
+         * {@linkplain net.minecraft.client.GraphicsStatus#FAST fast graphics}.
+         *
+         * @see #renderTypes(net.minecraftforge.client.RenderTypeGroup)
+         */
+        public SimpleBakedModel.Builder renderTypes(net.minecraftforge.client.RenderTypeGroup renderTypes, net.minecraftforge.client.RenderTypeGroup renderTypesFast) {
+            this.renderTypes = renderTypes;
+            this.renderTypesFast = renderTypesFast;
+            return this;
+        }
+
         public BakedModel build() {
             if (this.particleIcon == null) {
                 throw new RuntimeException("Missing particle!");
             } else {
                 Map<Direction, List<BakedQuad>> map = Maps.transformValues(this.culledFaces, ImmutableList.Builder::build);
                 return new SimpleBakedModel(
-                    this.unculledFaces.build(), new EnumMap<>(map), this.hasAmbientOcclusion, this.usesBlockLight, this.isGui3d, this.particleIcon, this.transforms
+                    // Forge: Account for render types in model JSONs
+                    this.unculledFaces.build(), new EnumMap<>(map), this.hasAmbientOcclusion, this.usesBlockLight, this.isGui3d, this.particleIcon, this.transforms, this.renderTypes, this.renderTypesFast
                 );
             }
         }
