--- a/net/minecraft/client/gui/screens/ChatScreen.java
+++ b/net/minecraft/client/gui/screens/ChatScreen.java
@@ -115,6 +_,7 @@
         } else if (p_426273_.isConfirmation()) {
             this.handleChatInput(this.input.getValue(), true);
             this.exitReason = ChatScreen.ExitReason.DONE;
+            if (this.minecraft.screen == this) // FORGE: Prevent closing the screen if another screen has been opened.
             this.minecraft.setScreen(null);
             return true;
         } else {
