--- a/net/minecraft/world/item/MobBucketItem.java
+++ b/net/minecraft/world/item/MobBucketItem.java
@@ -18,13 +_,18 @@
 import net.minecraft.world.level.material.Fluid;
 
 public class MobBucketItem extends BucketItem {
-    private final EntityType<? extends Mob> type;
-    private final SoundEvent emptySound;
+    private final java.util.function.Supplier<? extends EntityType<? extends Mob>> entityTypeSupplier;
+    private final java.util.function.Supplier<? extends SoundEvent> emptySoundSupplier;
 
+    @Deprecated
     public MobBucketItem(EntityType<? extends Mob> p_151137_, Fluid p_151138_, SoundEvent p_151139_, Item.Properties p_151140_) {
-        super(p_151138_, p_151140_);
-        this.type = p_151137_;
-        this.emptySound = p_151139_;
+        this(() -> p_151137_, () -> p_151138_, () -> p_151139_, p_151140_);
+    }
+
+    public MobBucketItem(java.util.function.Supplier<? extends EntityType<? extends Mob>> entitySupplier, java.util.function.Supplier<? extends Fluid> fluidSupplier, java.util.function.Supplier<? extends SoundEvent> soundSupplier, Item.Properties properties) {
+        super(fluidSupplier, properties);
+        this.emptySoundSupplier = soundSupplier;
+        this.entityTypeSupplier = entitySupplier;
     }
 
     @Override
@@ -37,11 +_,11 @@
 
     @Override
     protected void playEmptySound(@Nullable LivingEntity p_391719_, LevelAccessor p_151152_, BlockPos p_151153_) {
-        p_151152_.playSound(p_391719_, p_151153_, this.emptySound, SoundSource.NEUTRAL, 1.0F, 1.0F);
+        p_151152_.playSound(p_391719_, p_151153_, this.getEmptySound(), SoundSource.NEUTRAL, 1.0F, 1.0F);
     }
 
     private void spawn(ServerLevel p_151142_, ItemStack p_151143_, BlockPos p_151144_) {
-        Mob mob = this.type.create(p_151142_, EntityType.createDefaultStackConfig(p_151142_, p_151143_, null), p_151144_, EntitySpawnReason.BUCKET, true, false);
+        Mob mob = this.getFishType().create(p_151142_, EntityType.createDefaultStackConfig(p_151142_, p_151143_, null), p_151144_, EntitySpawnReason.BUCKET, true, false);
         if (mob instanceof Bucketable bucketable) {
             CustomData customdata = p_151143_.getOrDefault(DataComponents.BUCKET_ENTITY_DATA, CustomData.EMPTY);
             bucketable.loadFromBucketTag(customdata.copyTag());
@@ -52,5 +_,13 @@
             p_151142_.addFreshEntityWithPassengers(mob);
             mob.playAmbientSound();
         }
+    }
+
+    protected EntityType<? extends Mob> getFishType() {
+        return entityTypeSupplier.get();
+    }
+
+    protected SoundEvent getEmptySound() {
+        return emptySoundSupplier.get();
     }
 }
