--- a/net/minecraft/client/renderer/entity/EntityRenderDispatcher.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderDispatcher.java
@@ -207,6 +_,14 @@
         return this.itemInHandRenderer;
     }
 
+    public Map<PlayerModelType, AvatarRenderer<AbstractClientPlayer>> getPlayerRenderers() {
+        return java.util.Collections.unmodifiableMap(playerRenderers);
+    }
+
+    public Map<PlayerModelType, AvatarRenderer<ClientMannequin>> getMannequinRenderers() {
+        return java.util.Collections.unmodifiableMap(mannequinRenderers);
+    }
+
     @Override
     public void onResourceManagerReload(ResourceManager p_174004_) {
         EntityRendererProvider.Context entityrendererprovider$context = new EntityRendererProvider.Context(
@@ -224,5 +_,6 @@
         this.renderers = EntityRenderers.createEntityRenderers(entityrendererprovider$context);
         this.playerRenderers = EntityRenderers.createAvatarRenderers(entityrendererprovider$context);
         this.mannequinRenderers = EntityRenderers.createAvatarRenderers(entityrendererprovider$context);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onGatherLayers(renderers, playerRenderers, mannequinRenderers, entityrendererprovider$context);
     }
 }
