--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -95,6 +_,7 @@
         if (p_430553_.mapId != null) {
             int j = p_430553_.rotation % 4 * 2;
             p_431616_.mulPose(Axis.ZP.rotationDegrees(j * 360.0F / 8.0F));
+            if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderItemInFrame(p_430553_, this, p_431616_, p_423714_, p_430553_.lightCoords)) {
             p_431616_.mulPose(Axis.ZP.rotationDegrees(180.0F));
             float f2 = 0.0078125F;
             p_431616_.scale(0.0078125F, 0.0078125F, 0.0078125F);
@@ -102,6 +_,7 @@
             p_431616_.translate(0.0F, 0.0F, -1.0F);
             int i = this.getLightCoords(p_430553_.isGlowFrame, 15728850, p_430553_.lightCoords);
             this.mapRenderer.render(p_430553_.mapRenderState, p_431616_, p_423714_, true, i);
+            }
         } else if (!p_430553_.item.isEmpty()) {
             p_431616_.mulPose(Axis.ZP.rotationDegrees(p_430553_.rotation * 360.0F / 8.0F));
             int k = this.getLightCoords(p_430553_.isGlowFrame, 15728880, p_430553_.lightCoords);
