--- a/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -75,6 +_,7 @@
     }
 
     public void render(S p_364280_, PoseStack p_115311_, MultiBufferSource p_115312_, int p_115313_) {
+        if (net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderLivingPre(p_364280_, this, p_115311_, p_115312_, p_115313_)) return;
         p_115311_.pushPose();
         if (p_364280_.hasPose(Pose.SLEEPING)) {
             Direction direction = p_364280_.bedOrientation;
@@ -110,6 +_,7 @@
 
         p_115311_.popPose();
         super.render(p_364280_, p_115311_, p_115312_, p_115313_);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderLivingPost(p_364280_, this, p_115311_, p_115312_, p_115313_);
     }
 
     protected boolean shouldRenderLayers(S p_360804_) {
@@ -291,6 +_,7 @@
         p_363057_.isFullyFrozen = p_368665_.isFullyFrozen();
         p_363057_.isBaby = p_368665_.isBaby();
         p_363057_.isInWater = p_368665_.isInWater();
+        p_363057_.isInWater = p_368665_.isInWaterOrSwimmable();
         p_363057_.isAutoSpinAttack = p_368665_.isAutoSpinAttack();
         p_363057_.hasRedOverlay = p_368665_.hurtTime > 0 || p_368665_.deathTime > 0;
         ItemStack itemstack = p_368665_.getItemBySlot(EquipmentSlot.HEAD);
@@ -315,6 +_,7 @@
     }
 
     protected void extractAdditionalHitboxes(T p_392957_, Builder<HitboxRenderState> p_396936_, float p_397822_) {
+        super.extractAdditionalHitboxes(p_392957_, p_396936_, p_397822_);
         AABB aabb = p_392957_.getBoundingBox();
         float f = 0.01F;
         HitboxRenderState hitboxrenderstate = new HitboxRenderState(
