--- a/net/minecraft/client/renderer/block/model/BlockModel.java
+++ b/net/minecraft/client/renderer/block/model/BlockModel.java
@@ -28,8 +_,20 @@
     @Nullable Boolean ambientOcclusion,
     @Nullable ItemTransforms transforms,
     TextureSlots.Data textureSlots,
-    @Nullable ResourceLocation parent
+    @Nullable ResourceLocation parent,
+    @Nullable net.minecraftforge.client.model.ForgeBlockModelData forgeData
 ) implements UnbakedModel {
+    public BlockModel(
+        @Nullable UnbakedGeometry geometry,
+        @Nullable UnbakedModel.GuiLight guiLight,
+        @Nullable Boolean ambientOcclusion,
+        @Nullable ItemTransforms transforms,
+        TextureSlots.Data textureSlots,
+        @Nullable ResourceLocation parent
+    ) {
+        this(geometry, guiLight, ambientOcclusion, transforms, textureSlots, parent, null);
+    }
+
     @VisibleForTesting
     static final Gson GSON = new GsonBuilder()
         .registerTypeAdapter(BlockModel.class, new BlockModel.Deserializer())
@@ -97,8 +_,9 @@
                 unbakedmodel$guilight = UnbakedModel.GuiLight.getByName(GsonHelper.getAsString(jsonobject, "gui_light"));
             }
 
+            var forgeData = net.minecraftforge.client.ForgeHooksClient.deserializeBlockModel(jsonobject, p_111500_);
             ResourceLocation resourcelocation = s.isEmpty() ? null : ResourceLocation.parse(s);
-            return new BlockModel(unbakedgeometry, unbakedmodel$guilight, obool, itemtransforms, textureslots$data, resourcelocation);
+            return new BlockModel(unbakedgeometry, unbakedmodel$guilight, obool, itemtransforms, textureslots$data, resourcelocation, forgeData);
         }
 
         private TextureSlots.Data getTextureMap(JsonObject p_111510_) {
@@ -121,6 +_,9 @@
 
         @Nullable
         protected UnbakedGeometry getElements(JsonDeserializationContext p_111507_, JsonObject p_111508_) {
+            var geo = net.minecraftforge.client.ForgeHooksClient.deserializeBlockModelGeometry(p_111508_, p_111507_);
+            if (geo != null)
+                return geo;
             if (!p_111508_.has("elements")) {
                 return null;
             } else {
