--- a/net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
@@ -171,6 +_,7 @@
         WorldDataConfiguration worlddataconfiguration = SharedConstants.IS_RUNNING_IN_IDE
             ? new WorldDataConfiguration(new DataPackConfig(List.of("vanilla", "tests"), List.of()), FeatureFlags.DEFAULT_FLAGS)
             : WorldDataConfiguration.DEFAULT;
+        net.minecraftforge.event.ForgeEventFactory.addPackFindersServer(packrepository::addPackFinder);
         WorldLoader.InitConfig worldloader$initconfig = createDefaultLoadConfig(packrepository, worlddataconfiguration);
         CompletableFuture<WorldCreationContext> completablefuture = WorldLoader.load(
             worldloader$initconfig,
@@ -471,7 +_,7 @@
                                         if (p_269627_) {
                                             p_270552_.accept(this.uiState.getSettings().dataConfiguration());
                                         } else {
-                                            p_270552_.accept(WorldDataConfiguration.DEFAULT);
+                                            p_270552_.accept(new WorldDataConfiguration(new DataPackConfig(ImmutableList.of("vanilla"), ImmutableList.of()), FeatureFlags.DEFAULT_FLAGS)); // FORGE: Revert to *actual* vanilla data
                                         }
                                     },
                                     Component.translatable("dataPack.validation.failed"),
@@ -585,6 +_,7 @@
         if (path != null) {
             if (this.tempDataPackRepository == null) {
                 this.tempDataPackRepository = ServerPacksSource.createPackRepository(path, this.packValidator);
+                net.minecraftforge.resource.ResourcePackLoader.loadResourcePacks(this.tempDataPackRepository, false);
                 this.tempDataPackRepository.reload();
             }
 
