--- a/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java
@@ -105,6 +_,7 @@
         int i = this.leftPos;
         int j = this.topPos;
         super.render(p_409971_, p_409213_, p_408205_, p_408282_);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onContainerRenderBackground(this, p_409971_, p_409213_, p_408205_);
         p_409971_.pose().pushMatrix();
         p_409971_.pose().translate(i, j);
         this.renderLabels(p_409971_, p_409213_, p_408205_);
@@ -117,6 +_,8 @@
             this.onStopHovering(slot);
         }
 
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onContainerRenderForeground(this, p_409971_, p_409213_, p_408205_);
+
         p_409971_.pose().popMatrix();
     }
 
@@ -200,7 +_,7 @@
             ItemStack itemstack = this.hoveredSlot.getItem();
             if (this.menu.getCarried().isEmpty() || this.showTooltipWithItemInHand(itemstack)) {
                 p_283594_.setTooltipForNextFrame(
-                    this.font, this.getTooltipFromContainerItem(itemstack), itemstack.getTooltipImage(), p_282171_, p_281909_, itemstack.get(DataComponents.TOOLTIP_STYLE)
+                    this.font, this.getTooltipFromContainerItem(itemstack), itemstack.getTooltipImage(), itemstack, p_282171_, p_281909_
                 );
             }
         }
@@ -216,7 +_,8 @@
 
     private void renderFloatingItem(GuiGraphics p_282567_, ItemStack p_281330_, int p_281772_, int p_281689_, @Nullable String p_282568_) {
         p_282567_.renderItem(p_281330_, p_281772_, p_281689_);
-        p_282567_.renderItemDecorations(this.font, p_281330_, p_281772_, p_281689_ - (this.draggingItem.isEmpty() ? 0 : 8), p_282568_);
+        var font = net.minecraftforge.client.extensions.common.IClientItemExtensions.of(p_281330_).getFont(p_281330_, net.minecraftforge.client.extensions.common.IClientItemExtensions.FontContext.ITEM_COUNT);
+        p_282567_.renderItemDecorations(font == null ? this.font : font, p_281330_, p_281772_, p_281689_ - (this.draggingItem.isEmpty() ? 0 : 8), p_282568_);
     }
 
     protected void renderLabels(GuiGraphics p_281635_, int p_282681_, int p_283686_) {
@@ -317,7 +_,8 @@
         if (super.mouseClicked(p_427253_, p_427622_)) {
             return true;
         } else {
-            boolean flag = this.minecraft.options.keyPickItem.matchesMouse(p_427253_) && this.minecraft.player.hasInfiniteMaterials();
+            var mouseKey = com.mojang.blaze3d.platform.InputConstants.Type.MOUSE.getOrCreate(p_427253_.button());
+            boolean flag = this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey) && this.minecraft.player.hasInfiniteMaterials();
             Slot slot = this.getHoveredSlot(p_427253_.x(), p_427253_.y());
             this.doubleclick = this.lastClickSlot == slot && p_427622_;
             this.skipNextRelease = false;
@@ -327,6 +_,7 @@
                 int i = this.leftPos;
                 int j = this.topPos;
                 boolean flag1 = this.hasClickedOutside(p_427253_.x(), p_427253_.y(), i, j);
+                if (slot != null) flag1 = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
                 int k = -1;
                 if (slot != null) {
                     k = slot.index;
@@ -352,7 +_,7 @@
                         }
                     } else if (!this.isQuickCrafting) {
                         if (this.menu.getCarried().isEmpty()) {
-                            if (flag) {
+                            if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                                 this.slotClicked(slot, k, p_427253_.button(), ClickType.CLONE);
                             } else {
                                 boolean flag2 = k != -999 && p_427253_.hasShiftDown();
@@ -376,7 +_,7 @@
                                 this.quickCraftingType = 0;
                             } else if (p_427253_.button() == 1) {
                                 this.quickCraftingType = 1;
-                            } else if (flag) {
+                            } else if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                                 this.quickCraftingType = 2;
                             }
                         }
@@ -453,10 +_,13 @@
 
     @Override
     public boolean mouseReleased(MouseButtonEvent p_430939_) {
+        super.mouseReleased(p_430939_); //Forge, Call parent to release buttons
         Slot slot = this.getHoveredSlot(p_430939_.x(), p_430939_.y());
         int i = this.leftPos;
         int j = this.topPos;
         boolean flag = this.hasClickedOutside(p_430939_.x(), p_430939_.y(), i, j);
+        if (slot != null) flag = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
+        var mouseKey = com.mojang.blaze3d.platform.InputConstants.Type.MOUSE.getOrCreate(p_430939_.button());
         int k = -1;
         if (slot != null) {
             k = slot.index;
@@ -473,7 +_,7 @@
                         if (slot2 != null
                             && slot2.mayPickup(this.minecraft.player)
                             && slot2.hasItem()
-                            && slot2.container == slot.container
+                            && slot2.isSameInventory(slot)
                             && AbstractContainerMenu.canItemQuickReplace(slot2, this.lastQuickMoved, true)) {
                             this.slotClicked(slot2, slot2.index, p_430939_.button(), ClickType.QUICK_MOVE);
                         }
@@ -538,7 +_,7 @@
 
                 this.slotClicked(null, -999, AbstractContainerMenu.getQuickcraftMask(2, this.quickCraftingType), ClickType.QUICK_CRAFT);
             } else if (!this.menu.getCarried().isEmpty()) {
-                if (this.minecraft.options.keyPickItem.matchesMouse(p_430939_)) {
+                if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                     this.slotClicked(slot, k, p_430939_.button(), ClickType.CLONE);
                 } else {
                     boolean flag1 = k != -999 && p_430939_.hasShiftDown();
@@ -609,7 +_,7 @@
     public boolean keyPressed(KeyEvent p_424324_) {
         if (super.keyPressed(p_424324_)) {
             return true;
-        } else if (this.minecraft.options.keyInventory.matches(p_424324_)) {
+        } else if (this.minecraft.options.keyInventory.isActiveAndMatches(com.mojang.blaze3d.platform.InputConstants.getKey(p_424324_))) {
             this.onClose();
             return true;
         } else {
@@ -688,6 +_,13 @@
 
         super.onClose();
     }
+
+    @org.jetbrains.annotations.Nullable
+    public Slot getSlotUnderMouse() { return this.hoveredSlot; }
+    public int getGuiLeft() { return leftPos; }
+    public int getGuiTop() { return topPos; }
+    public int getXSize() { return imageWidth; }
+    public int getYSize() { return imageHeight; }
 
     @OnlyIn(Dist.CLIENT)
     record SnapbackData(ItemStack item, Vector2i start, Vector2i end, long time) {
