--- a/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
@@ -65,6 +_,9 @@
                 framesize = new FrameSize(nativeimage.getWidth(), nativeimage.getHeight());
             }
 
+            SpriteContents contents = net.minecraftforge.client.ForgeHooksClient.loadSpriteContents(p_421043_, p_421044_, framesize, nativeimage, list);
+            if (contents != null) return contents;
+
             return new SpriteContents(p_421043_, framesize, nativeimage, optional, list);
         };
     }
