--- a/net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
@@ -51,6 +_,7 @@
         Equippable equippable = p_366444_.get(DataComponents.EQUIPPABLE);
         if (equippable != null && shouldRender(equippable, p_117122_)) {
             A a = this.getArmorModel(p_423438_, p_117122_);
+            var model = this.getArmorModel(p_423438_, p_117122_, p_366444_, a);
             EquipmentClientInfo.LayerType equipmentclientinfo$layertype = this.usesInnerModel(p_117122_)
                 ? EquipmentClientInfo.LayerType.HUMANOID_LEGGINGS
                 : EquipmentClientInfo.LayerType.HUMANOID;
@@ -58,7 +_,7 @@
                 .renderLayers(
                     equipmentclientinfo$layertype,
                     equippable.assetId().orElseThrow(),
-                    a,
+                    model,
                     p_423438_,
                     p_366444_,
                     p_117119_,
@@ -75,5 +_,12 @@
 
     private boolean usesInnerModel(EquipmentSlot p_117129_) {
         return p_117129_ == EquipmentSlot.LEGS;
+    }
+
+    /**
+     * Hook to allow item-sensitive armor model. for HumanoidArmorLayer.
+     */
+    protected net.minecraft.client.model.Model getArmorModel(S state, EquipmentSlot slot, ItemStack itemStack, A _default) {
+       return net.minecraftforge.client.ForgeHooksClient.getArmorModel(state, itemStack, slot, _default);
     }
 }
