--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -233,6 +_,9 @@
             case null:
             default:
                 this.clearPostEffect();
+                var rl = p_109107_ == null ? null : net.minecraftforge.client.EntitySpectatorShaderManager.get(p_109107_.getType());
+                if (rl != null)
+                    this.setPostEffect(rl);
         }
     }
 
@@ -244,7 +_,13 @@
     public void processBlurEffect() {
         PostChain postchain = this.minecraft.getShaderManager().getPostChain(BLUR_POST_CHAIN_ID, LevelTargetBundle.MAIN_TARGETS);
         if (postchain != null) {
+            // TODO: [Forge][Rendering] Check if this blend management is needed anymnore with the new render pipeline changes
+            boolean wasBlendEnabled = com.mojang.blaze3d.opengl.GlStateManager._isBlendEnabled();
+            if (wasBlendEnabled)
+                com.mojang.blaze3d.opengl.GlStateManager._disableBlend();
             postchain.process(this.minecraft.getMainRenderTarget(), this.resourcePool);
+            if (wasBlendEnabled)
+                com.mojang.blaze3d.opengl.GlStateManager._enableBlend();
         }
     }
 
@@ -406,7 +_,7 @@
                 f *= Mth.lerp(f2, 1.0F, 0.85714287F);
             }
 
-            return f;
+            return net.minecraftforge.client.event.ForgeEventFactoryClient.fireComputeFov(this, p_109142_, p_109143_, f, p_109144_).getFOV();
         }
     }
 
@@ -557,7 +_,7 @@
                 }
             } else if (flag && this.minecraft.screen != null) {
                 try {
-                    this.minecraft.screen.renderWithTooltipAndSubtitles(guigraphics, i, j, p_343467_.getGameTimeDeltaTicks());
+                    net.minecraftforge.client.ForgeHooksClient.drawScreen(this.minecraft.screen, guigraphics, i, j, p_343467_.getRealtimeDeltaTicks());
                 } catch (Throwable throwable1) {
                     CrashReport crashreport1 = CrashReport.forThrowable(throwable1, "Rendering screen");
                     CrashReportCategory crashreportcategory1 = crashreport1.addCategory("Screen render details");
@@ -672,6 +_,10 @@
                         BlockInWorld blockinworld = new BlockInWorld(this.minecraft.level, blockpos, false);
                         Registry<Block> registry = this.minecraft.level.registryAccess().lookupOrThrow(Registries.BLOCK);
                         flag = !itemstack.isEmpty() && (itemstack.canBreakBlockInAdventureMode(blockinworld) || itemstack.canPlaceOnBlockInAdventureMode(blockinworld));
+                        if (!flag && blockstate.hasBlockEntity()) {
+                            var blockEntity = this.minecraft.level.getBlockEntity(blockpos);
+                            flag = blockEntity != null && blockEntity.hasCustomOutlineRendering();
+                        }
                     }
                 }
             }
@@ -706,6 +_,9 @@
         if (this.minecraft.options.bobView().get()) {
             this.bobView(posestack, this.mainCamera.getPartialTickTime());
         }
+
+        var cameraSetup = net.minecraftforge.client.event.ForgeEventFactoryClient.fireComputeCameraAngles(this, this.mainCamera, f);
+        this.mainCamera.setRotation(cameraSetup.getYaw(), cameraSetup.getPitch(), cameraSetup.getRoll());
 
         matrix4f.mul(posestack.last().pose());
         float f3 = this.minecraft.options.screenEffectScale().get().floatValue();
