--- a/com/mojang/blaze3d/opengl/GlTexture.java
+++ b/com/mojang/blaze3d/opengl/GlTexture.java
@@ -18,8 +_,13 @@
     protected boolean modesDirty = true;
 
     protected GlTexture(String p_393950_, TextureFormat p_392837_, int p_394590_, int p_391379_, int p_391947_, int p_396659_) {
+        this(p_393950_, p_392837_, p_394590_, p_391379_, p_391947_, p_396659_, false);
+    }
+
+    protected GlTexture(String p_393950_, TextureFormat p_392837_, int p_394590_, int p_391379_, int p_391947_, int p_396659_, boolean stencil) {
         super(p_393950_, p_392837_, p_394590_, p_391379_, p_391947_);
         this.id = p_396659_;
+        this.stencilEnabled = stencil;
     }
 
     @Override
@@ -86,5 +_,15 @@
     public void setTextureFilter(FilterMode p_394977_, FilterMode p_397175_, boolean p_393558_) {
         super.setTextureFilter(p_394977_, p_397175_, p_393558_);
         this.modesDirty = true;
+    }
+
+
+    /* FORGE START */
+
+    private final boolean stencilEnabled;
+
+    @Override
+    public boolean isStencilEnabled() {
+        return this.stencilEnabled;
     }
 }
