--- a/net/minecraft/client/renderer/special/SpecialModelRenderers.java
+++ b/net/minecraft/client/renderer/special/SpecialModelRenderers.java
@@ -199,6 +_,8 @@
             map.put(Blocks.TRAPPED_CHEST, GIFT_CHEST);
         }
 
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onCreateSpecialBlockRenderers(map);
+
         Builder<Block, SpecialModelRenderer<?>> builder = ImmutableMap.builder();
         map.forEach((p_421033_, p_421034_) -> {
             SpecialModelRenderer<?> specialmodelrenderer = p_421034_.bake(p_431302_);
